### PR TITLE
PT-2073: Save button accessibility

### PR DESCRIPTION
**Pull request by ErickLipin**
https://virtocommerce.atlassian.net/browse/PT-174

My changes are from 37 to 76 rows. Another changes are made by CodeMaid extension text formatting.

### CODE REVIEW

File: src/VirtoCommerce.MarketingModule.Web/Scripts/dynamicContent/blades/items/item-detail.js [hunk 1]
```diff
@@ -1,173 +1,198 @@
...
+            };
 
-            if (!blade.isNew) {
-               
+            blade.IsChanged = function () {
+                // Changes check method ensures the fact of different presentation way of dynamic properties:
+                // In the whole object and dynamic properties thru va-generic-value-input template directive
+                // First check objects for full equivalence
+                let equals = angular.equals(blade.origEntity, blade.currentEntity);
+                if (!equals) {
```
Comment (Vectorfield4) [inline src/VirtoCommerce.MarketingModule.Web/Scripts/dynamicContent/blades/items/item-detail.js:38]: Inverse it.

File: src/VirtoCommerce.MarketingModule.Web/Scripts/dynamicContent/blades/items/item-detail.js [hunk 2]
```diff
@@ -1,173 +1,198 @@
...
-                            return blade.IsChanged() && $scope.formScope.$valid;
-                        },
-                        permission: blade.updatePermission
-                    },
-                    {
-                        name: "platform.commands.reset", icon: 'fa fa-undo',
-                        executeMethod: function () {
-                            angular.copy(blade.origEntity, blade.currentEntity);
+                    if (blade.origEntity.dynamicProperties && blade.currentEntity.dynamicProperties &&
+                        blade.origEntity.dynamicProperties.length && blade.currentEntity.dynamicProperties.length) {
```
Comment (Vectorfield4) [inline src/VirtoCommerce.MarketingModule.Web/Scripts/dynamicContent/blades/items/item-detail.js:43]: Inverse it.

**Review by Vectorfield4 — CHANGES_REQUESTED** [review]
https://betterprogramming.pub/refactoring-guard-clauses-2ceeaa1a9da

**Review by akak1977 — APPROVED** [review]

File: src/VirtoCommerce.MarketingModule.Web/Scripts/dynamicContent/blades/items/item-detail.js [hunk 3]
```diff
@@ -1,173 +1,197 @@
...
-                result = angular.equals(_.omit(blade.origEntity, ['dynamicProperties']), _.omit(blade.currentEntity, ['dynamicProperties']));
-                if (blade.origEntity.dynamicProperties && blade.currentEntity.dynamicProperties &&
-                    blade.origEntity.dynamicProperties.length > 0 && blade.currentEntity.dynamicProperties.length > 0 &&
-                    blade.origEntity.dynamicProperties[0].values.length > 0 && blade.currentEntity.dynamicProperties[0].values.length > 0 ) {
-                    result = result && blade.origEntity.dynamicProperties[0].values[0].valueId == blade.currentEntity.dynamicProperties[0].values[0].value.id;
+                equals = angular.equals(_.omit(blade.origEntity, ['dynamicProperties']), _.omit(blade.currentEntity, ['dynamicProperties']));
+                if (!blade.origEntity.dynamicProperties || !blade.currentEntity.dynamicProperties ||
+                    !blade.origEntity.dynamicProperties.length || !blade.currentEntity.dynamicProperties.length) return !equals;
+
+                for (var origEntityDynamicProperty of blade.origEntity.dynamicProperties) {
```
Comment (ksavosteev) [inline src/VirtoCommerce.MarketingModule.Web/Scripts/dynamicContent/blades/items/item-detail.js:45]: originalEntityDynamicProperty

File: src/VirtoCommerce.MarketingModule.Web/Scripts/dynamicContent/blades/items/item-detail.js [hunk 4]
```diff
@@ -1,173 +1,197 @@
...
+                    blade.initialize();
+                }
+            };
+
+            blade.IsChanged = function () {
+                // Changes check method ensures the fact of different presentation way of dynamic properties:
+                // In the whole object and dynamic properties thru va-generic-value-input template directive
+                // First check objects for full equivalence
+                let equals = angular.equals(blade.origEntity, blade.currentEntity);
+                if (equals) return !equals;
```
Comment (ksavosteev) [inline src/VirtoCommerce.MarketingModule.Web/Scripts/dynamicContent/blades/items/item-detail.js:38]: return false
move to a new line

File: src/VirtoCommerce.MarketingModule.Web/Scripts/dynamicContent/blades/items/item-detail.js [hunk 5]
```diff
@@ -1,173 +1,197 @@
...
+
                 // Check possibility they are really equivalent but have a different representation of dynamic property value
-                result = angular.equals(_.omit(blade.origEntity, ['dynamicProperties']), _.omit(blade.currentEntity, ['dynamicProperties']));
-                if (blade.origEntity.dynamicProperties && blade.currentEntity.dynamicProperties &&
-                    blade.origEntity.dynamicProperties.length > 0 && blade.currentEntity.dynamicProperties.length > 0 &&
-                    blade.origEntity.dynamicProperties[0].values.length > 0 && blade.currentEntity.dynamicProperties[0].values.length > 0 ) {
-                    result = result && blade.origEntity.dynamicProperties[0].values[0].valueId == blade.currentEntity.dynamicProperties[0].values[0].value.id;
+                equals = angular.equals(_.omit(blade.origEntity, ['dynamicProperties']), _.omit(blade.currentEntity, ['dynamicProperties']));
+                if (!blade.origEntity.dynamicProperties || !blade.currentEntity.dynamicProperties ||
+                    !blade.origEntity.dynamicProperties.length || !blade.currentEntity.dynamicProperties.length) return !equals;
```
Comment (ksavosteev) [inline src/VirtoCommerce.MarketingModule.Web/Scripts/dynamicContent/blades/items/item-detail.js:43]: return false
move to a new line

File: src/VirtoCommerce.MarketingModule.Web/Scripts/dynamicContent/blades/items/item-detail.js [hunk 6]
```diff
@@ -1,173 +1,197 @@
...
+
+                for (var origEntityDynamicProperty of blade.origEntity.dynamicProperties) {
+                    let currEntityDynamicProperty = blade.currentEntity.dynamicProperties.find(x => x.name == origEntityDynamicProperty.name);
```
Comment (ksavosteev) [inline src/VirtoCommerce.MarketingModule.Web/Scripts/dynamicContent/blades/items/item-detail.js:46]: currentEntityDynamicProperty

File: src/VirtoCommerce.MarketingModule.Web/Scripts/dynamicContent/blades/items/item-detail.js [hunk 7]
```diff
@@ -1,173 +1,197 @@
...
+                        // 'Every' func require length check
+                        equals = origEntityDynamicProperty.values.length
+                            ? origEntityDynamicProperty.values.every(x => currDictionaryValuesIds.includes(x.valueId))
+                            : !currDictionaryValuesIds.length;
+                    }
+                    else if (origEntityDynamicProperty.isArray &&
+                        origEntityDynamicProperty.values.length && currEntityDynamicProperty.values.length) {
+                        // arrays comparison
+                        // Check arrays equality (adding an element after the same element deletation)
+                        let origValues = origEntityDynamicProperty.values.map(x => x.value);
```
Comment (ksavosteev) [inline src/VirtoCommerce.MarketingModule.Web/Scripts/dynamicContent/blades/items/item-detail.js:62]: originalValues

File: src/VirtoCommerce.MarketingModule.Web/Scripts/dynamicContent/blades/items/item-detail.js [hunk 8]
```diff
@@ -1,173 +1,197 @@
...
+                        let currValues = currEntityDynamicProperty.values.map(x => x.value);
```
Comment (ksavosteev) [inline src/VirtoCommerce.MarketingModule.Web/Scripts/dynamicContent/blades/items/item-detail.js:63]: currentValues

File: src/VirtoCommerce.MarketingModule.Web/Scripts/dynamicContent/blades/items/item-detail.js [hunk 9]
```diff
@@ -1,173 +1,197 @@
...
+                        equals = origEntityDynamicProperty.values.length == currEntityDynamicProperty.values.length && origValues.every(x => currValues.includes(x));
+                    } else {
+                        // simple types comparison
+                        // !!AND EMPTY ARRAYS!!
```
Comment (ksavosteev) [inline src/VirtoCommerce.MarketingModule.Web/Scripts/dynamicContent/blades/items/item-detail.js:67]: remove or make the comment more clear

File: src/VirtoCommerce.MarketingModule.Web/Scripts/dynamicContent/blades/items/item-detail.js [hunk 10]
```diff
@@ -1,173 +1,197 @@
...
+                    if (!equals)
+                        break;
                 }
-            }
-            return !result;
-        };
-       
-        blade.initialize = function () {
-            blade.toolbarCommands = [];
+                return !equals;
```
Comment (ksavosteev) [inline src/VirtoCommerce.MarketingModule.Web/Scripts/dynamicContent/blades/items/item-detail.js:74]: add an empty line above

File: src/VirtoCommerce.MarketingModule.Web/Scripts/dynamicContent/blades/items/item-detail.js (hunk 4 above)
Comment (Vectorfield4) [inline src/VirtoCommerce.MarketingModule.Web/Scripts/dynamicContent/blades/items/item-detail.js:38]:
```suggestion
                if (equals)
                     return false;
                     
```

File: src/VirtoCommerce.MarketingModule.Web/Scripts/dynamicContent/blades/items/item-detail.js (hunk 5 above)
Comment (Vectorfield4) [inline src/VirtoCommerce.MarketingModule.Web/Scripts/dynamicContent/blades/items/item-detail.js:43]:
```suggestion
                if ( !blade.origEntity.dynamicProperties ||
                     !blade.currentEntity.dynamicProperties ||
                     !blade.origEntity.dynamicProperties.length ||
                     !blade.currentEntity.dynamicProperties.length ) 
                return !equals;
                
```

**Review by Vectorfield4 — CHANGES_REQUESTED** [review]

**Review by ksavosteev — APPROVED** [review]

**Review by Vectorfield4 — APPROVED** [review]

**Pull request merged**